File: services/api-gateway/src/common/mtls.ts
import { Agent } from 'https';
import { readFileSync } from 'fs';

// Node counterpart to services/mpc-party/mtls.go's clientTLSConfigFromEnv and
// services/temporal-worker/activities/mtls.go -- same three-env-var opt-in
// convention (MTLS_CERT_FILE / MTLS_KEY_FILE / MTLS_CA_FILE), same certs, since
// both sides are configured from the same Vault-PKI-issued bundle in practice
// (infrastructure/terraform/modules/vault-pki, issued per-pod by
// services/vault-pki-init).
//
// Precedence, matching the Go implementations exactly:
//   - none of the three set: mTLS is off, plain HTTP. A valid, if less secure,
//     local-dev configuration -- not an error.
//   - all three set: an https.Agent presenting this service's client
//     certificate and trusting only the given CA.
//   - partially set, or files unreadable: throws. A half-configured mTLS setup
//     silently falling back to plaintext is the failure mode worth being loud
//     about -- the same reason NewActivities calls log.Fatalf on this case.
export const MTLS_CERT_FILE = 'MTLS_CERT_FILE';
export const MTLS_KEY_FILE = 'MTLS_KEY_FILE';
export const MTLS_CA_FILE = 'MTLS_CA_FILE';

export interface MtlsAgentResult {
  agent: Agent | undefined;
  enabled: boolean;
}

export function mtlsAgentFromEnv(env: NodeJS.ProcessEnv = process.env): MtlsAgentResult {
  const certFile = env[MTLS_CERT_FILE];
  const keyFile = env[MTLS_KEY_FILE];
  const caFile = env[MTLS_CA_FILE];

  const set = [certFile, keyFile, caFile].filter(Boolean).length;
  if (set === 0) {
    return { agent: undefined, enabled: false };
  }
  if (set !== 3) {
    throw new Error(
      `mTLS is partially configured: set all of ${MTLS_CERT_FILE}, ${MTLS_KEY_FILE}, ${MTLS_CA_FILE}, or none of them`,
    );
  }

  let cert: Buffer;
  let key: Buffer;
  let ca: Buffer;
  try {
    cert = readFileSync(certFile as string);
    key = readFileSync(keyFile as string);
    ca = readFileSync(caFile as string);
  } catch (err) {
    throw new Error(`failed to read mTLS material: ${(err as Error).message}`);
  }

  return {
    agent: new Agent({
      cert,
      key,
      ca,
      // Explicit rather than relying on the default: the CA above is the only
      // trust anchor for these internal links, and the server's certificate
      // must genuinely chain to it.
      rejectUnauthorized: true,
    }),
    enabled: true,
  };
}

// Axios/HttpModule config fragment. Returns {} when mTLS isn't configured so
// callers can spread it unconditionally.
export function mtlsHttpOptions(env: NodeJS.ProcessEnv = process.env): { httpsAgent?: Agent } {
  const { agent } = mtlsAgentFromEnv(env);
  return agent ? { httpsAgent: agent } : {};
}

File: services/temporal-worker/activities/mtls.go
package activities

import (
	"crypto/tls"
	"os"
)

// Client side of mutual TLS to services/mpc-party -- see the matching
// doc comment in services/mpc-party/mtls.go for why this link specifically
// (the DKG ceremony transport) and how a real deployment issues these
// certs via Vault PKI. Same three-env-var opt-in convention as the server
// side, and same names, since both are configured from the same
// Terraform/Vault-issued cert bundle in practice.
const (
	envMTLSCertFile = "MTLS_CERT_FILE"
	envMTLSKeyFile  = "MTLS_KEY_FILE"
	envMTLSCAFile   = "MTLS_CA_FILE"
)

// sharedReloader is the process's single view of its mTLS material, so the
// worker cannot end up with two loaders disagreeing about which generation
// of its certificate is current.
var sharedReloader *clientCertReloader

// clientTLSConfigFromEnv mirrors serverTLSConfigFromEnv in
// services/mpc-party/mtls.go: (nil, false, nil) when mTLS isn't
// configured (a valid, if less secure, local-dev configuration), an error
// only when it's partially configured or the files don't load.
//
// The returned config follows certificate rotation -- see certreload.go for
// why loading the keypair exactly once means the worker stops being able to
// reach any party the moment its startup certificate expires.
func clientTLSConfigFromEnv() (*tls.Config, bool, error) {
	certFile, keyFile, caFile := os.Getenv(envMTLSCertFile), os.Getenv(envMTLSKeyFile), os.Getenv(envMTLSCAFile)
	if certFile == "" || keyFile == "" || caFile == "" {
		return nil, false, nil
	}

	if sharedReloader == nil {
		r, err := newClientCertReloader(certFile, keyFile, caFile)
		if err != nil {
			return nil, false, err
		}
		sharedReloader = r
	}
	return sharedReloader.config(), true, nil
}
